src/jit/morph.cpp

```diff
@@ -8698,6 +8698,44 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
         }
     }
 
+    // Optimize Math.Pow(x, c)
+    if ((call->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC) &&
+        info.compCompHnd->getIntrinsicID(call->gtCallMethHnd) == CORINFO_INTRINSIC_Pow)
+    {
+        noway_assert(call->fgArgInfo->ArgCount() == 2);
+        GenTree* arg0    = gtArgEntryByArgNum(call, 0)->node;
+        GenTree* arg1    = gtArgEntryByArgNum(call, 1)->node;
+        GenTree* newNode = nullptr;
+
+        if (!arg0->IsCnsFltOrDbl() && arg1->IsCnsFltOrDbl())
+        {
+            GenTreeDblCon* powerCon = arg1->AsDblCon();
+            if (powerCon->gtDconVal == 2.0) // or should I compare int64 with 0x4000000000000000 ?
+            {
+                // Math.Pow(x, 2) -> x*x
+                newNode = gtNewOperNode(GT_MUL, powerCon->TypeGet(), arg0,
+                                        gtNewLclvNode(arg0->gtLclVar.gtLclNum, arg0->TypeGet()));
```

Comment:
It seems that you are assuming that `arg0` is `GT_LCL_VAR`. Why would that always be true?

Reply:
@mikedn So it can be GT_LCL_VAR, GT_CALL, GT_FIELD (+ static field access with initialization). I've temporarily limited it to LCL_VAR, I guess for other cases I need to introduce a tmp var?

Reply:
> So it can be GT_LCL_VAR, GT_CALL, GT_FIELD

I don't know why it would be limited to those opers, it can probably be pretty much anything, like `a + b * (float)c`. 

Anyway, yes, if it's not `GT_LCL_VAR` (or `GT_LCL_FLD` perhaps) you need to introduce a temporary variable because the original expression now has multiple uses in `x * x`. Introducing a temp is a bit unfortunate because it breaks trees and that may cause other problems. In the past I toyed with the idea of adding a `SQR` oper to avoid multiple uses though I'm not sure if it's something worth doing.

Reply:
yeah no idea how to handle cases like 
```csharp
return cond1() && cond2() && Math.Pow(MethodWithSideEffects(), 2.0);
```
I've modified the code to handle both local variables (GT_LCL_VAR) and fields (it comes as GT_IND) cases.
Also, `Math.Pow(x, 1) and Math.Pow(x, -1) should fine for any kind of arg0 (don't have to clone it)